org.opent2t.sample.lamp.superpopular/com.wink.lightbulb/js/thingTranslator.js

```diff
@@ -1,110 +1,145 @@
+/* jshint esversion: 6 */
+/* jshint node: true */
 'use strict';
+const WinkHelper = require('opent2t-translator-helper-wink');
+
+// This code uses ES2015 syntax that requires at least Node.js v4.
+// For Node.js ES2015 support details, reference http://node.green/
+
+function validateArgumentType(arg, argName, expectedType) {
+    if (typeof arg === 'undefined') {
+        throw new Error('Missing argument: ' + argName + '. ' +
+            'Expected type: ' + expectedType + '.');
+    } else if (typeof arg !== expectedType) {
+        throw new Error('Invalid argument: ' + argName + '. ' +
+            'Expected type: ' + expectedType + ', got: ' + (typeof arg));
+    }
+}
 
-var https = require('https');
+// Wink does not always populate every desired_state property, but last_reading doesn't necessarily
+// update as soon as we send our PUT request. Instead of relying just on one state or the other,
+// we use this StateReader class to read from desired_state if it is there, and fall back to last_reading
+// if it is not.
+class StateReader {
+    constructor(desired_state, last_reading) {
+        this.desired_state = desired_state;
+        this.last_reading = last_reading;
+    }
 
-// logs device state
-function logDeviceState(device) {
-    if (typeof (device) !== 'undefined') {
-        console.log('  device.name          : ' + device.name);
-        console.log('  device.props         : ' + device.props);
-    } else {
-        console.log('device is undefined');
+    get(state) {
+        if (this.desired_state[state] !== undefined) {
+            return this.desired_state[state];
+        }
+        else {
+            return this.last_reading[state];
+        }
     }
-};
+}
 
-// Helper method to send power state commands to Wink
-function sendPowerStateCommandToDevice(powerState) {
+// Helper method to convert the device schema to the translator schema.
+function deviceSchemaToTranslatorSchema(deviceSchema) {
+    var stateReader = new StateReader(deviceSchema.desired_state, deviceSchema.last_reading);
 
-    var postData = JSON.stringify({
-        'desired_state': {
-            'powered': powerState
-        }
-    });
-    var options = {
-        protocol: 'https:',
-        host: 'api.wink.com',
-        path: '/light_bulbs/' + deviceId,
-        headers: {
-            'Authorization': 'Bearer ' + accessToken,
-            'Content-Type': 'application/json',
-            'Accept': 'application/json',
-            'Content-length': postData.length
-        },
-        method: 'PUT'
+    var powered = stateReader.get('powered');
+
+    return {
+        id: deviceSchema['object_type'] + '.' + deviceSchema['object_id'],
+        n: deviceSchema['name'],
+        rt: 'org.opent2t.sample.thermostat.superpopular',
+        power: { 'value': powered }
     };
```

Comment:
Yeah, I'll look into doing this as part of a separate PR.

Not sure if I have a Wink bulb that supports color, but I should be able to find one at least that supports dimming.

---
In reply to: [76681179](https://github.com/openT2T/translators/pull/84#discussion_r76681179) [](ancestors = 76681179)